inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/region_yolo_gpu_ref.cl

```diff
@@ -12,94 +12,95 @@
-            // coords: x/y
-            int index = FUNC_CALL(entry_index)(IW, IH, COORDS, CLASSES, INPUTS_COUNT, b, n * IW * IH, 0);
-            int i = index + 2 * x;
-            output[i] = FUNC_CALL(logistic_activate)(input[i]);
-            output[i+1] = FUNC_CALL(logistic_activate)(input[i+1]);
+    in_i = INPUT0_GET_INDEX(batch, 1 + region_offset, y_index, x_index);
+    out_i = FUNC_CALL(output_index)(batch, region_num, xy, 1);
+    output[out_i] = FUNC_CALL(logistic_activate)(input[in_i]);
 
-            // coords: w/h: directly copy?
-            index = FUNC_CALL(entry_index)(IW, IH, COORDS, CLASSES, INPUTS_COUNT, b, n * IW * IH, 2);
-            i = index + 2 * x;
-            output[i] = input[i];
-            output[i+1] = input[i+1];
+    /// width,height
+    in_i = INPUT0_GET_INDEX(batch, 2 + region_offset, y_index, x_index);
+    out_i = FUNC_CALL(output_index)(batch, region_num, xy, 2);
+    output[out_i] = input[in_i];
 
-            // confidence
-            index = FUNC_CALL(entry_index)(IW, IH, COORDS, CLASSES, INPUTS_COUNT, b, n * IW * IH, COORDS);
-            for (int j = 0; j < CONF_CLASSES; j++)
-            {
-                i = index + x + j*IH*IW;
-                output[i] = FUNC_CALL(logistic_activate)(input[i]);
-            }
-        }
-    }
+    in_i = INPUT0_GET_INDEX(batch, 3 + region_offset, y_index, x_index);
+    out_i = FUNC_CALL(output_index)(batch, region_num, xy, 3);
+    output[out_i] = input[in_i];
 
+    /// objectness score
+    in_i = INPUT0_GET_INDEX(batch, COORDS + region_offset, y_index, x_index);
+    out_i = FUNC_CALL(output_index)(batch, region_num, xy, COORDS);
+    output[out_i] = FUNC_CALL(logistic_activate)(input[in_i]);
+
+    /// class score(confidence)
 #if DO_SOFTMAX
-    // the probability of classes
-    int index = FUNC_CALL(entry_index)(IW, IH, COORDS, CLASSES, INPUTS_COUNT, 0, 0, COORDS + 1);
-    int batch_offset = INPUTS_COUNT / NUM;
-    for (int b = 0; b < IB * NUM; b++)
-        FUNC_CALL(softmax_generic)(input + index + b * batch_offset, output + index + b * batch_offset,
-                                   1, CLASSES, IH, IW, x);
+    in_i = INPUT0_GET_INDEX(batch, COORDS + 1 + region_offset, y_index, x_index);
+    UNIT_TYPE max_value = input[in_i];
+    for (int j = 1; j < CLASSES; j++) {
+        in_i = INPUT0_GET_INDEX(batch, COORDS + 1 + j + region_offset, y_index, x_index);
+        max_value = max(max_value, input[in_i]);
+    }
+
+    UNIT_TYPE expSum = 0;
+    for (int j = 0; j < CLASSES; j++) {
+        in_i = INPUT0_GET_INDEX(batch, COORDS + 1 + j + region_offset, y_index, x_index);
+        out_i = FUNC_CALL(output_index)(batch, region_num, xy, COORDS + 1 + j);
+        output[out_i] = exp(input[in_i] - max_value);
+        expSum += output[out_i];
+    }
+
+    for (int j = 0; j < CLASSES; j++) {
+        out_i = FUNC_CALL(output_index)(batch, region_num, xy, COORDS + 1 + j);
+        output[out_i] /= expSum;
+    }
```

Comment:
Could you check this performance on realistic input size? This accesses global memory repeatedly, hope to check the performance.

Reply:
ok, I'll check it.